### ISSUE TITLE
Use Class-File API for tests on JDK 24+

### ISSUE THREAD

**Issue by wmdietl** [eisop/checker-framework#1198]
The [Class-File API](https://openjdk.org/jeps/484) was finalized in JDK 24, but the old implementation continued to work in JDK 24.

In JDK 25, some nonjunit tests fail because the old implementation was removed.
See #1190 for failures like below.

We should guard these jtreg tests by a version check and add a new implementation based on the Class-File API for JDK 24+.

````
direct:
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/../PersistUtil.java:5: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.ClassFile;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/../PersistUtil.java:37: error: cannot find symbol
    public static ClassFile compileAndReturn(String fullFile, String testClass) throws Exception {
                  ^
  symbol:   class ClassFile
  location: class PersistUtil
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/Driver.java:4: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.Annotation;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/Driver.java:5: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.ClassFile;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:5: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.Annotation;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:6: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.Attribute;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:7: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.ClassFile;
                              ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:8: error: package com.sun.tools.classfile.ConstantPool does not exist
import com.sun.tools.classfile.ConstantPool.InvalidIndex;
                                           ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:9: error: package com.sun.tools.classfile.ConstantPool does not exist
import com.sun.tools.classfile.ConstantPool.UnexpectedEntry;
                                           ^
/home/runner/work/checker-framework/checker-framework/checker/jtreg/nullness/inheritDeclAnnoPersist/ReferenceInfoUtil.java:10: error: package com.sun.tools.classfile does not exist
import com.sun.tools.classfile.Method;
````

**Comment by wmdietl** [comment]
@thisisalexandercook Could you take a look into this, as you're already looking into the Class-File API?

**Comment by thisisalexandercook** [comment]
@wmdietl I have been looking for ways to transition the jtreg test harness to only use the new classfile API when running tests on bytecode. One option we previously discussed was somehow compiling the tests to specific bytecode releases but then running them through a  JDK 25+ harness in order to utilize the new API. This may be a possible solution:

Using the javac `--source` and `--target` flags, we can setup a jtreg test as follows

```java
/*
 * @test
 * @summary Constructors default‑qualifier – class‑file version XX (Java xx)
 * @requires jdk.version.major >= 25
 *
 * @compile --source xx --target xx  Constructors.java ...
 * @run main Driver Constructors
 */

```

Its my understanding that the bytecode should be compiled to the java xx (i.e. 11, 15 etc..) standard in this configuration but new APIs remain visible so we won't get compilation errors when importing new JDK features (this is unlike using the `--release` flag, see [JEP 247](https://openjdk.org/jeps/247)). With this setup, our tests would be restricted to running when $JAVA_HOME = JDK 25+ so that the harness could use the new API, but would be compiled to whatever bytecode version we set. We could then have test stubs for each desired bytecode version that just have a jtreg compilation setup (as shown above), and reference a single test case and new classfile API harness. 

If this seems viable, we would probably want to wait until JDK 25 has an official release to implement this. In the meantime, I could push a PR that keeps the old API with a separate 25+ API implementation, and open a new issue for what is described above.
